### PR TITLE
Unset values in predefined VM

### PR DESCRIPTION
**Pull request by ermau**

### CODE REVIEW

File: Xamarin.PropertyEditing.Tests/PredefinedValuesViewModelTests.cs [hunk 1]
```diff
@@ -196,6 +268,76 @@ internal enum PredefinedEnumTest
...
+			vm.ValueName = "test";
+
+			var info = editor.values[property.Object] as ValueInfo<string>;
+			Assert.That (info, Is.Not.Null);
+			Assert.That (info.ValueDescriptor, Is.EqualTo ("test"));
+			Assert.That (info.Source, Is.EqualTo (ValueSource.Local));
+		}
+
+		[Test]
+		[Description ("When passing along strings in the value desciprtor, we should ensure they are empty and not null for differentiation")]
```
Comment (bleroy) [inline Xamarin.PropertyEditing.Tests/PredefinedValuesViewModelTests.cs]: Same typo :D

File: Xamarin.PropertyEditing/ViewModels/PredefinedValuesViewModel.cs [hunk 2]
```diff
@@ -16,42 +15,34 @@ public PredefinedValuesViewModel (TargetPlatform platform, IPropertyInfo propert
 			if (this.predefinedValues == null)
 				throw new ArgumentException (nameof(property) + " did not have predefined values", nameof(property));
 
+			var list = new List<string> (this.predefinedValues.PredefinedValues.Keys);
+			// If we're constrained but can't use the default, we need a blank to represent default
+			if (this.predefinedValues.IsConstrainedToPredefined) {
+				if (!property.ValueSources.HasFlag (ValueSources.Default) || !TryGetValueName (default(TValue), out string defaultName)) {
```
Comment (bleroy) [inline Xamarin.PropertyEditing/ViewModels/PredefinedValuesViewModel.cs]: Curious about why these aren't be combined with `&&`. Is it the pattern matching?

File: Xamarin.PropertyEditing/ViewModels/PropertyViewModel.cs [hunk 3]
```diff
@@ -75,12 +75,21 @@ protected virtual void OnValueChanged ()
...
-					else if (currentValue.Source != valueInfo.Source || !Equals (currentValue.Value, valueInfo.Value)) {
-						// Even if the value is the same, they are not equal if the source is not the same because
-						// it means the value is set differently at the source.
-						disagree = true;
-						currentValue = null;
-						break;
+					else {
+						if (currentValue.Source != valueInfo.Source) {
+							currentValue.Source = ValueSource.Default;
+							disagree = true;
```
Comment (bleroy) [inline Xamarin.PropertyEditing/ViewModels/PropertyViewModel.cs]: nit: you can move `disagree = true` out of the `if`s.

File: Xamarin.PropertyEditing.Tests/MockObjectEditor.cs [hunk 4]
```diff
@@ -180,9 +181,14 @@ public Task<IReadOnlyList<ITypeInfo>> GetAssignableTypesAsync (IPropertyInfo pro
...
-					return info;
-				else if (value == null || value is T) {
+				if (info != null) {
+					return new ValueInfo<T> {
+						CustomExpression = info.CustomExpression,
+						Source = info.Source,
+						ValueDescriptor = info.ValueDescriptor,
+						Value = info.Value
+					};
+				} else if (value == null || value is T) {
```
[the remark this answers is not in the document]
Comment (ermau) [reply Xamarin.PropertyEditing.Tests/MockObjectEditor.cs]: I'm behind on some of the newer pattern matching stuff, so I'm just going to stick with what I know and is clear to me for the moment.

[the remark this answers is not in the document]
Comment (ermau) [reply Xamarin.PropertyEditing.Tests/MockObjectEditor.cs]: Oh actually I see what you're talking about here, I didn't realize I was casting value down there, yeah we can change that, I do know about that bit.

**Pull request merged**